Disable WordPress object cache on post count WP_Query call

jon-heller opened the pull request:
We were receiving a PHP error regarding the memory size exhausted when attempting to run a full reindex (of a large database, over 5000 posts and hundreds of taxonomies). This error occurred when running this function both through the WP admin as well as wp-cli. 

We limited the number of synchronous indexes using the `algolia_indexing_batch_size` filter as suggested, even setting that to 1, but still received the error. We also noticed the error always occurred almost immediately after starting the reindex. 

After some PHP debugging, we found the function was crashing at line 30 of WordPress core file `cache.php`. This was the result of the Algolia function `get_re_index_items_count`. So it was exhausting memory just trying to count the posts. 

I’m not sure, but it didn’t seem necessary or even helpful (?) to cache the results during a reindex. So we disabled the WordPress object caching in the WP Query call of the above function:

```php
		$query = new WP_Query( array(
			'post_type'   		    => $this->post_types,
			'post_status' 		    => 'any', // Let the `should_index` take care of the filtering.
			'suppress_filters'      => true,
			'cache_results'  => false
		) );
```

This fixed the error and the indexing ran without a problem. 

There's another WP_Query call in that file to get the actual posts. For some reason this doesn't cause a memory error, but maybe it makes sense to disable cache there as well?

Hoping that this can be incorporated into the code so that we don’t have to use a modified version of the plugin with this line.

Thanks!

Mamaduka commented:
I think improve this even more with something like this -
 https://github.com/10up/ElasticPress/blob/develop/bin/wp-cli.php#L795.

jon-heller commented:
I encountered more memory issues in our live environment (hosted by Pantheon), and had to add these additional parameters to the query as well:

```
      'lazy_load_term_meta' => false,
      'update_post_term_cache' => false
```

rayrutjes commented:
Thanks a lot for that feedback @jon-heller.
Disabling cache makes a lot of sense in the context of re-indexing.

Regarding the 2 additional parameters: `lazy_load_term_meta ` & `update_post_term_cache `, do you know if this could lead to backward compatibility issues?

I'd gladly move forward with this ASAP.

I'll also need to take a look at @Mamaduka's link.

PS: I'm OoO for 2 weeks, please expect some delay in my answers ;)

jon-heller commented:
Great, glad to hear this seems to make sense.

As for those parameters, `lazy_load_term_meta` is newer and I think was introduced in 4.6. I _think_ that WP_Query will just ignore any parameters it doesn't recognize though.

Mamaduka commented:
I wrote small WP-CLI [script](https://gist.github.com/Mamaduka/00222f1873f9fc49c565710ca95701af) to do simple benchmarking using Symfony [Stopwatch](https://symfony.com/doc/current/components/stopwatch.html) component.

Tested on local WP installation with 5000+ posts and dozens of taxos. I got significant memory reduction after adding WP_CLI's [`Utils\wp_clear_object_cache()`](https://github.com/wp-cli/wp-cli/blob/master/php/utils-wp.php#L226) into the loop. Setting query cache params, didn't show almost any memory reduction on average. Of course results will be different, base on setup and size of DB.

Test results:
```
Before WP_CLI\Utils\wp_clear_object_cache(): 154.00 MiB - 6622 ms
After WP_CLI\Utils\wp_clear_object_cache(): 66.00 MiB - 5818 ms
```

P.S. @rayrutjes we can use WP-CLI `Utils\wp_clear_object_cache()` instead of the code I linked above.

rayrutjes commented:
Thanks for the investigation @Mamaduka ,

Will have a look when back to the office.

rayrutjes commented:
Hi @jon-heller , I'm willing to move forward here.
I think we can go with @Mamaduka 's suggestion for the wp-cli command.

Then for the non wp-cli, we also need to add the `lazy_load_term_meta` and `update_post_term_cache` parameters maybe. 

Do you think you can add those?

jon-heller commented:
@rayrutjes No problem, I added the additional two options to `get_re_index_items_count`. I also added the same parameters to `get_items`, since it seems the same logic would apply here as well.

rayrutjes commented:
This has been merged here: #633 ,

Thanks a lot for all your efforst @jon-heller  and @Mamaduka 🚀 

Let me know if you have any further suggestions.
